libcudacxx/include/cuda/std/detail/libcxx/include/__string

```diff
@@ -0,0 +1,1099 @@
+{
+    size_t __len = 0;
+    for (; !eq(*__s, char_type(0)); ++__s)
+        ++__len;
+    return __len;
+}
+
+template <class _CharT>
+inline
+_CCCL_CONSTEXPR_CXX14 const _CharT*
+char_traits<_CharT>::find(const char_type* __s, size_t __n, const char_type& __a)
+{
+    for (; __n; --__n)
+    {
+        if (eq(*__s, __a))
+            return __s;
+        ++__s;
+    }
+    return 0;
+}
+
+template <class _CharT>
+inline
+_LIBCUDACXX_INLINE_VISIBILITY _CharT*
+char_traits<_CharT>::move(char_type* __s1, const char_type* __s2, size_t __n)
+{
+    char_type* __r = __s1;
+    if (__s1 < __s2)
+    {
+        for (; __n; --__n, ++__s1, ++__s2)
+            assign(*__s1, *__s2);
+    }
+    else if (__s2 < __s1)
+    {
+        __s1 += __n;
+        __s2 += __n;
+        for (; __n; --__n)
+            assign(*--__s1, *--__s2);
+    }
+    return __r;
+}
+
+template <class _CharT>
+inline
+_LIBCUDACXX_INLINE_VISIBILITY _CharT*
+char_traits<_CharT>::copy(char_type* __s1, const char_type* __s2, size_t __n)
+{
+    _LIBCUDACXX_ASSERT(__s2 < __s1 || __s2 >= __s1+__n, "char_traits::copy overlapped range");
+    char_type* __r = __s1;
+    for (; __n; --__n, ++__s1, ++__s2)
+        assign(*__s1, *__s2);
+    return __r;
+}
+
+template <class _CharT>
+inline
+_LIBCUDACXX_INLINE_VISIBILITY _CharT*
+char_traits<_CharT>::assign(char_type* __s, size_t __n, char_type __a)
+{
+    char_type* __r = __s;
+    for (; __n; --__n, ++__s)
+        assign(*__s, __a);
+    return __r;
+}
+
+// char_traits<char>
+
+// GCC's builtin_strlen isn't reliable at constexpr time
+// MSVC does not expose builtin_strlen before C++17
+#if defined(_CCCL_COMPILER_GCC) || (defined(_CCCL_COMPILER_MSVC) && _CCCL_STD_VER < 2017)
+#define _CCCL_HAS_NO_BUILTIN_STRLEN
+#endif
+
+template <>
+struct _LIBCUDACXX_TEMPLATE_VIS char_traits<char>
+{
+    typedef char      char_type;
+    typedef int       int_type;
+    typedef streamoff off_type;
+#ifndef _LIBCUDACXX_HAS_NO_WCHAR_H
+    typedef streampos pos_type;
+    typedef mbstate_t state_type;
+#endif // !_LIBCUDACXX_HAS_NO_WCHAR_H
+
+    _LIBCUDACXX_INLINE_VISIBILITY static inline _CCCL_CONSTEXPR_CXX14
+    void assign(char_type& __c1, const char_type& __c2) noexcept {__c1 = __c2;}
+    _LIBCUDACXX_INLINE_VISIBILITY static inline constexpr bool eq(char_type __c1, char_type __c2) noexcept
+            {return __c1 == __c2;}
+    _LIBCUDACXX_INLINE_VISIBILITY static inline constexpr bool lt(char_type __c1, char_type __c2) noexcept
+        {return (unsigned char)__c1 < (unsigned char)__c2;}
+
+    _LIBCUDACXX_INLINE_VISIBILITY static _CCCL_CONSTEXPR_CXX14
+    int compare(const char_type* __s1, const char_type* __s2, size_t __n) noexcept;
+    _LIBCUDACXX_INLINE_VISIBILITY static inline size_t _CCCL_CONSTEXPR_CXX14
+    length(const char_type* __s)  noexcept {
+#ifdef _CCCL_HAS_NO_BUILTIN_STRLEN
+#ifdef _LIBCUDACXX_IS_CONSTANT_EVALUATED // is_constant_evaluated only exists since GCC 9
+        if (__libcpp_is_constant_evaluated())
+#endif
```

Comment:
Important: We *should* be able to just use `__libcpp_is_constant_evaluated()` everywhere

Reply:
No, the logic here is such that I need this path to be taken _despite_ the builtin not existing, which makes the wrapper return `false`. In general you are right, but this is a special case working around several things at once.